stan/math/prim/mat/fun/dogleg.hpp

```diff
@@ -0,0 +1,83 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_DOGLEG_HPP
+#define STAN_MATH_PRIM_MAT_FUN_DOGLEG_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <unsupported/Eigen/NonLinearOptimization>
+
+namespace stan {
+  namespace math {
+
+    template <typename _Scalar, int NX = Eigen::Dynamic,
+      int NY = Eigen::Dynamic>
+    struct NLOFunctor {
+      typedef _Scalar Scalar;
+      enum {
+        InputsAtCompileTime = NX,
+        ValuesAtCompileTime = NY
+      };
+
+      typedef Eigen::Matrix<Scalar, InputsAtCompileTime, 1> InputType;
+      typedef Eigen::Matrix<Scalar, ValuesAtCompileTime, 1> ValueType;
+      typedef Eigen::Matrix<Scalar, ValuesAtCompileTime, InputsAtCompileTime>
+        JacobianType;
+
+      const int m_inputs, m_values;
+
+      NLOFunctor() : m_inputs(InputsAtCompileTime),
+        m_values(ValuesAtCompileTime) {}
+      NLOFunctor(int inputs, int values) : m_inputs(inputs), m_values(values) {}
+
+      int inputs() const { return m_inputs; }
+      int values() const { return m_values; }
+    };
+
+    template <typename F1, typename F2>
+    struct hybrj_functor : NLOFunctor<double> {
+    private:
+      F1 f1;
+      F2 f2;
+
+    public:
+      hybrj_functor(const F1& f1_param, const F2& f2_param) {
+        f1 = f1_param;
+        f2 = f2_param;
+      }
+
+
+      int operator()(const Eigen::VectorXd &x, Eigen::VectorXd &fvec) {
```

Comment:
I followed closely Eigen's example implementation of the dogleg solver and (wrongfully) overlooked a lot of the details. I've been able to implement your previous comments in `dogleg.hpp`, but this one causes the compiler some grievance.

The functor is used in Eigen's unsupported module, which checks the value the functor returns. See `eigen_3.2.9/unsupported/Eigen/src/NonLinearOptimization/HybridNonLinearSolver.h`:
```
if ( functor(x, fvec) < 0) return HybridNonLinearSolverSpace::UserAsked;
```

I need to check  exactly what this line does. I'm guessing Eigen uses return types to identify different kinds of functors, some of which get used for the dogleg algebraic solver, others for other nonlinear optimization methods (such as function minimization).